src/soc/intel/baytrail/txei.c

```diff
@@ -0,0 +1,1196 @@
+		printk(BIOS_INFO, "TXE requested a reset\n");
+		system_reset();
+		break;
+	case PWR_CYCLE_RESET:
+		printk(BIOS_INFO, "TXE requested a power cycle reset\n");
+		full_reset();
+		break;
+	case PERFORM_GLOBAL_RESET:
+		printk(BIOS_INFO, "TXE requested a global reset\n");
+		printk(BIOS_INFO, "Global reset will be performed in ramstage\n");
+		break;
+	default:
+		printk(BIOS_INFO, "TXE invalid BIOS action %x\n",
+		       bios_action >> DID_ACK_DATA_SHIFT);
+		break;
+	}
+}
+
+/*
+ * Initialize the TXE device with provided temporary BAR. If BAR is 0 use a
+ * default. This is intended for pre-mem usage only where BARs haven't been
+ * assigned yet and devices are not enabled.
+ */
+void txe_txei_init(void)
+{
+	pci_devfn_t dev = PCI_DEV(0, TXE_DEV, TXE_FUNC);
+	uint32_t aliveness_req, aliveness_resp;
+	uint32_t txe_opmode;
+	uintptr_t tempbar0, tempbar;
+	struct stopwatch sw;
+
+	/* Check if device enabled */
+	if (!is_txe_devfn_visible())
+		return;
+
+	tempbar0 =  pci_read_config32(dev, PCI_BASE_ADDRESS_0);
+	tempbar0 &= ~PCI_BASE_ADDRESS_MEM_ATTR_MASK;
+	tempbar =  pci_read_config32(dev, PCI_BASE_ADDRESS_1);
+	tempbar &= ~PCI_BASE_ADDRESS_MEM_ATTR_MASK;
+
+	txe_opmode = pci_read_config32(dev, TXE_FWSTS0);
+	txe_opmode &= TXE_CURR_OPMODE_MASK;
+	txe_opmode >>= TXE_CURR_OPMODE_SHIFT;
+
+	if (txe_opmode == TXE_FWSTS0_COM_DEBUG ||
+	    txe_opmode == TXE_FWSTS0_COM_SOFT_TEMP_DISABLE ||
+	    txe_opmode == TXE_FWSTS0_COM_SECOVER_JMPR ||
+	    txe_opmode == TXE_FWSTS0_COM_SECOVER_TXEI_MSG)
+		return;
+
+	/* Assign TXEI resource and enable the resource */
+	txei_assign_resource(dev, tempbar0, tempbar);
+
+	/* Configure interrupt */
+	pci_and_config8(dev, TXE_MSI_CAP, 0xfc);
+
+	stopwatch_init_usecs_expire(&sw, HECI_READY_TIMEOUT);
+	do {
+		aliveness_req = read32p(tempbar + TXE_MMIO_ALIVENESS_REQ) & 1;
+		aliveness_resp = read32p(tempbar + TXE_MMIO_ALIVENESS_RESP) & 1;
+		if (stopwatch_expired(&sw)) {
+			printk(BIOS_ERR, "Timeout waiting for TXE aliveness early sync\n");
+			return;
+		}
+
+		udelay(1000);
+	} while(aliveness_req != aliveness_resp);
+
+	if ((read32p(tempbar + TXE_MMIO_ALIVENESS_REQ) & 1) == 1)
+		txe_aliveness_request(tempbar, false);
+
+	write32p(tempbar + TXE_MMIO_HOST_IPC_READINESS,
+		 read32p(tempbar + TXE_MMIO_HOST_IPC_READINESS) | IPC_READINESS_RDY_CLR);
+
+	stopwatch_init_usecs_expire(&sw, HECI_READY_TIMEOUT);
+	do {
+		if (stopwatch_expired(&sw)) {
+			printk(BIOS_ERR, "Timeout waiting for TXE IPC readiness\n");
+			return;
+		}
+
+		udelay(1000);
+	} while(!(read32p(tempbar + TXE_MMIO_SEC_IPC_READINESS) & IPC_READINESS_SEC_RDY));
+
+	write32p(tempbar + TXE_MMIO_HHISR,
+		 read32p(tempbar + TXE_MMIO_HHISR) & ~(HHISR_INT_BAR0_STS | HHISR_INT_BAR0_STS));
+
+	write32p(tempbar + TXE_MMIO_SEC_IPC_OUTPUT_STATUS,
+		 read32p(tempbar + TXE_MMIO_SEC_IPC_OUTPUT_STATUS) | IPC_OUTPUT_READY);
+
+
+	write32p(tempbar + TXE_MMIO_HOST_IPC_READINESS,
+		 read32p(tempbar + TXE_MMIO_HOST_IPC_READINESS) | IPC_READINESS_HOST_RDY);
```

Comment:
This could use some comments.

Reply:
I would add them if I knew what it really does and what for... I just barely managed to get it working...